Source/ExecutionEngine/ImplementationTask.cs

```diff
@@ -1,35 +1,97 @@
+using System;
 using System.IO;
+using System.Reactive.Subjects;
 using System.Threading;
 using System.Threading.Tasks;
+using VC;
 
 namespace Microsoft.Boogie;
 
 public interface IImplementationTask {
-  public Task<VerificationResult> ActualTask { get; }
+  IObservable<VerificationStatus> ObservableStatus { get; }
+  VerificationStatus CurrentStatus { get; }
+  ProcessedProgram ProcessedProgram { get; }
+  Implementation Implementation { get; }
+  Task<VerificationResult> ActualTask { get; }
   void Run();
 }
 
 public class ImplementationTask : IImplementationTask {
-  private CancellationTokenSource source;
-  private readonly ExecutionEngine engine;
-  private readonly ProcessedProgram processedProgram;
+  private readonly CancellationTokenSource taskCancellationSource;
+
+  private readonly Subject<VerificationStatus> observableStatus = new();
+  
+  private VerificationStatus currentStatus;
+  public IObservable<VerificationStatus> ObservableStatus => observableStatus;
+
+  public VerificationStatus CurrentStatus {
+    get => currentStatus;
+    private set {
+      currentStatus = value;
+      observableStatus.OnNext(value);
+    }
+  }
+  public ProcessedProgram ProcessedProgram { get; }
 
-  public Task<VerificationResult> ActualTask { get; private set; }
+  public Task<VerificationResult> ActualTask { get; }
   public Implementation Implementation { get; }
 
+  private readonly TaskCompletionSource runSource = new();
+  
   public ImplementationTask(ExecutionEngine engine, ProcessedProgram processedProgram, Implementation implementation) {
-    this.engine = engine;
-    this.processedProgram = processedProgram;
+    ProcessedProgram = processedProgram;
     Implementation = implementation;
+    taskCancellationSource = new CancellationTokenSource();
+    
+    VerificationResult verificationResult = engine.GetCachedVerificationResult(Implementation, TextWriter.Null);
+    if (verificationResult != null) {
+      ActualTask = Task.FromResult(verificationResult);
+      CurrentStatus = StatusFromVerificationResult(verificationResult);
+    } else {
+      CurrentStatus = VerificationStatus.Stale;
+      ActualTask = runSource.Task.ContinueWith(t => {
+
+        var enqueueTask = engine.EnqueueVerifyImplementation(ProcessedProgram, new PipelineStatistics(), null, null,
+          Implementation, taskCancellationSource, TextWriter.Null);
+
+        if (enqueueTask.IsCompleted) {
+          CurrentStatus = VerificationStatus.Verifying;
+        } else {
+          CurrentStatus = VerificationStatus.Queued;
+          enqueueTask.ContinueWith(_ => { CurrentStatus = VerificationStatus.Verifying; });
+        }
```

Comment:
Could the task itself send notifications, rather than having the two layers of `Task`?